etw/cs/traceloggingdynamic/TraceLoggingDynamic.cs

```diff
@@ -1741,34 +1900,118 @@ private void AddScalarDataNulTerminatedByteString(byte[] value, int startIndex,
             this.data.SetUInt8(pos + copyLength, 0);
         }
 
-        private void AddScalarDataNulTerminatedString(String value, int startIndex, int count)
+        private unsafe void AddScalarDataNulTerminatedString(String value, int startIndex, int count)
         {
-            Debug.Assert(count >= 0);
-            Debug.Assert(count <= value.Length - startIndex);
+            int endIndex = value.IndexOf('\0', startIndex, count);
+            int copyLength = endIndex < 0
+                ? count
+                : endIndex - startIndex;
+
+            int valueSize = copyLength * sizeof(UInt16);
+            int pos = this.data.ReserveSpaceFor(valueSize + sizeof(UInt16));
+            fixed (char* pValue = value)

```

Comment:
Note: Using MemoryCopy instead of `Encoding.Unicode` is a moderate performance improvement.